### PR TITLE
Switches to correct Dalli class by version

### LINKED ISSUE #1777: Dalli::Server is deprecated, use Dalli::Protocol::Binary instead
Recent versions of [dalli](https://rubygems.org/gems/dalli) (v3.0.0+) have deprecated `Dalli::Server`. That class has been renamed to `Dalli::Protocol::Binary`, but otherwise I believe the interface is similar.

The usage of `Dalli::Server` in this library causes deprecation warning emissions at application start up, most likely from the following call:

https://github.com/DataDog/dd-trace-rb/blob/e391d2eb64d3c6151a4bdd2710c6a8c7c1d57457/lib/ddtrace/contrib/dalli/patcher.rb#L20-L22

### PR DESCRIPTION
**Pull request by kbacha**
In version 3.0.0 Dalli switched from Dalli::Server to
Dalli::Protocol::Binary. While the former still exists it is generating
a warning upon load.

This change handles the case when dalli is below 3.0.0 and when it is
above.

Fixes #1777

### CODE REVIEW

**Comment by menglish-taskrabbit** [comment]
👍

**Comment by kbacha** [comment]
I've been having some trouble with appraisal and getting it running locally, especially on the new M1 chips. Do I need to add a new appraisal group for dalli 3.0?

**Comment by ivoanjo** [comment]
> Do I need to add a new appraisal group for dalli 3.0?

Yes, that sounds good!

> I've been having some trouble with appraisal and getting it running locally, especially on the new M1 chips.

Not good! If you're up to it, would you mind opening a separate issue with for the M1 contributor experience issue? Afaik none of the team uses one yet, but we definitely would like to have a smooth dev experience for people on such a setup :)

**Comment by delner** [comment]
Code changes look fine, but yeah, we're gonna need to fix CI/appraisals first. Once that's done, I'm happy to approve this one.

**Comment by kbacha** [comment]
@ivoanjo @delner 

Ok cool, several things that I've taken care of that are worth noting:

1. I added `platform: linux/x86_64` to all the services in the docker-compose file. This enables it to build on M1 macbooks but does come with a cost penalty.
2. Added dalli 2.x to the `contrib-old` appraisal group. This seemed like a good fit and didn't require a new gemfile. It also made it easier to align with the ruby versions.
3. Made the appraisal config for dalli version specific. Dalli 3.x is incompatible with ruby 3.x and so I restricted the testing to those versions. For versions under ruby 3 I've specified dalli as being less than 3.0
4. Added `APPRAISAL_GROUP` environment variable. It would take hours to build the gemfiles on my laptop due to the performance penalty and the fact that I was building gemfiles unrelated to dalli. Setting this variable to a group name allowed me to target specific builds.
5. Updates all the appraisal gemfiles and gemfile.locks which have dalli listed in them.
6. Has the tests skip the check when the version is wrong. This prevents class errors in lower versions of ruby.

Thanks for the reviews! I look forward to additional feedback.

**Comment by kbacha** [comment]
Ooof, it looks like the gemfile's have gone out of date while i've waited for approval the past 2 weeks.

Could you take another look?? @ivoanjo @delner

**Comment by ivoanjo** [comment]
Hey Kevin! Apologies for the radio silence. Let me sync with the rest of the team and get back to you.

Don't bother chasing the gemfile conflicts -- we'll take care of updating them before merging so as to avoid repeated work trying to keep up with master 😄

File: Appraisals
Comment (marcotc) [inline Appraisals:17]: I suggest overriding the built-in `appraise` method here, in order to keep this file looking like a plain appraisal file going forward.

I suggest adding `alias appraise add_appraise` after you declare this method, and reverting the method name changes that follow up in this file.

**Review by marcotc — COMMENTED** [review]
Thank you so much again, @kbacha. I left one minor comment, but overall this approach is great and looks really flushed out! 🙇

File: Appraisals
Comment (kbacha) [reply Appraisals:17]: Ok, I think you mean to alias the original method and then define a new method called appraise. Since this file is evaled by appraisal I'm unsure if `super` would work with just a override so I'm trying the alias.

**Comment by ivoanjo** [comment]
Thank you very much @kbacha for the contribution! And double thanks for the patience dealing with the appraisals setup -- we take supporting older Rubies quite seriously, although I admit it can be a paaaaaaain sometimes.

I've rebased your work on top of current master, and will just wait for the final doublecheck from @marcotc to get this in!

**Review by marcotc — APPROVED** [review]

**Comment by marcotc** [comment]
Thank you so much, @kbacha 🙇!

**Pull request merged**
